x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java

```diff
@@ -0,0 +1,580 @@
+                geometry
+            );
+        } else {
+            return asGeometryDocValueReader(
+                new CartesianShapeCoordinateEncoder(),
+                new CartesianShapeIndexer("SpatialRelatesFunction"),
+                geometry
+            );
+        }
+
+    }
+
+    protected static GeometryDocValueReader asGeometryDocValueReader(
+        CoordinateEncoder encoder,
+        ShapeIndexer shapeIndexer,
+        Geometry geometry
+    ) throws IOException {
+        GeometryDocValueReader reader = new GeometryDocValueReader();
+        CentroidCalculator centroidCalculator = new CentroidCalculator();
+        if (geometry instanceof Circle circle) {
+            // TODO: How should we deal with circles?
+            centroidCalculator.add(new Point(circle.getX(), circle.getY()));
+        } else {
+            centroidCalculator.add(geometry);
+        }
+        reader.reset(GeometryDocValueWriter.write(shapeIndexer.indexShape(geometry), encoder, centroidCalculator));
+        return reader;
+    }
+
+    @Override
+    public boolean foldable() {
+        return left().foldable() && right().foldable();
+    }
+
+    @Override
+    public Object fold() {
+        return EvaluatorMapper.super.fold();
+    }
+
+    public abstract SpatialRelatesFunction withDocValues();
+
+    @Override
+    public int hashCode() {
+        // NB: the hashcode is currently used for key generation so
+        // to avoid clashes between aggs with the same arguments, add the class name as variation
+        return Objects.hash(getClass(), children(), useDocValues);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (super.equals(obj)) {
+            SpatialRelatesFunction other = (SpatialRelatesFunction) obj;
+            return Objects.equals(other.children(), children()) && Objects.equals(other.useDocValues, useDocValues);
+        }
+        return false;
+    }
+
+    public boolean useDocValues() {
+        return useDocValues;
+    }
+
+    /**
+     * Produce a map of rules defining combinations of incoming types to the evaluator factory that should be used.
+     */
+    protected abstract Map<SpatialEvaluatorKey, SpatialEvaluatorFactory<?, ?>> evaluatorRules();
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(
+        Function<Expression, EvalOperator.ExpressionEvaluator.Factory> toEvaluator
+    ) {
+        if (crsType == null) {
+            resolveType();
+        }
+        SpatialEvaluatorFactory<?, ?> factory = getSpatialEvaluatorFactory();
+        return factory.get(toEvaluator);
+    }
+
+    protected SpatialEvaluatorFactory<?, ?> getSpatialEvaluatorFactory() {
+        if (crsType == null) {
+            resolveType();
+        }
+        var evaluatorKey = new SpatialEvaluatorKey(crsType, useDocValues, fieldKey(left()), fieldKey(right()));
+        SpatialEvaluatorFactory<?, ?> factory = evaluatorRules().get(evaluatorKey);
```

Comment:
It'd be nice if the rules were a `static final` map rather than built every time. That's a lot of map entries.

Reply:
Yes, that was my plan. I think I forgot to take that last step. Will do so now.

Reply:
Actually, the current factory wrappers are not static inner classes, but make use of `source()` and `left()` and `right()` for convenience. Since this map is created at planning time, I do not think this should be too big an issue to create each time a query is planned. If you still think it is better to make this static, I can investigate a way to pass `this` into the class at planning time.

Reply:
OK. I've refactored and cleaned it up, so these are now all static.